datafusion/functions-aggregate/src/median.rs

```diff
@@ -230,6 +276,212 @@ impl<T: ArrowNumericType> Accumulator for MedianAccumulator<T> {
     }
 }
 
+/// The median groups accumulator accumulates the raw input values
+///
+/// For calculating the accurate medians of groups, we need to store all values
+/// of groups before final evaluation.
+/// So values in each group will be stored in a `Vec<T>`, and the total group values
+/// will be actually organized as a `Vec<Vec<T>>`.
+///
+#[derive(Debug)]
+struct MedianGroupsAccumulator<T: ArrowNumericType + Send> {
+    data_type: DataType,
+    group_values: Vec<Vec<T::Native>>,
```

Comment:
Just wonder -- using `Vec<Vec<>>` for as a state storage doesn't seem to differ much from a regular accumulator, but this PR still introduces a noticeable performance improvement. Are there any other optimizations that could be used in regular accumulator?

P.S. asking just because when I was doing +- same for count distinct ([PR](https://github.com/apache/datafusion/pull/8721)), the performance for GroupsAccumulator with `Vec<HashSet<>>` was not that significant comparing to regular accumulators with `HashSet<>` states.

Reply:
I think among other things, the intermediate state management (creating ListArrays directly rather than from ScalarValue) probably helps a lot: 

https://github.com/apache/datafusion/blob/6c9355d5be8b6045865fed67cb6d028b2dfc2e06/datafusion/functions-aggregate/src/median.rs#L200-L199

There is also an extra allocation per group when using the groups accumulator adapter thingie

That being said, it is a fair question how much better the existing MedianAccumulator could be if it built the ListArrays as does this PR directly 🤔

Reply:
@korowa I think what mentioned by @alamb is a important about the improvement.

Following are some other points for me:
- in `GroupsAccumulatorAdapter::update_batch`, we need to reorder the `input batch`, and use `slice` to split the reordered batch after. I think such two operations may be not cheap.
https://github.com/apache/datafusion/blob/6c9355d5be8b6045865fed67cb6d028b2dfc2e06/datafusion/functions-aggregate-common/src/aggregate/groups_accumulator.rs#L241-L265

- in `GroupsAccumulatorAdapter::merge_batch`, the similar problem as `input batch` may be even more serious... Becasue we need to reorder a `ListArray`

- and in `GroupsAccumulatorAdapter::state`, extra allocations exist as mentioned by @alamb .

Reply:
@korowa it means impl a `group accumulator` for `distinct count` not get a obviously improvement?
It is really surprise for me, I am learning https://github.com/apache/datafusion/pull/8721

Reply:
The re was some improvements, but overall results were like x2.63 for GroupsAccumulator, and x2.30 for the regular Accumulator -- so it would be like 13-15% overall difference, which is not as massive as this PR results.

However, maybe things has changed in GroupsAccumulator implementation, and now even plain `Vec<HashSet<>>` will be way faster.

Reply:
Seem really worth seeking the reason more deeply.